src/battle/ability.c

```diff
@@ -111,6 +111,93 @@ const u16 PowderMoveList[] = {
+        MOVE_ABSORB,
+        MOVE_DRAIN_PUNCH,
+        MOVE_DRAINING_KISS,
+        MOVE_DREAM_EATER,
+        MOVE_FLORAL_HEALING,
+        MOVE_GIGA_DRAIN,
+        MOVE_HEAL_ORDER,
+        MOVE_HEAL_PULSE,
+        MOVE_HEALING_WISH,
+        MOVE_HORN_LEECH,
+        MOVE_LEECH_LIFE,
+        MOVE_LUNAR_DANCE,
+        MOVE_MEGA_DRAIN,
+        MOVE_MILK_DRINK,
+        MOVE_MOONLIGHT,
+        MOVE_MORNING_SUN,
+        MOVE_OBLIVION_WING,
+        MOVE_PARABOLIC_CHARGE,
+        MOVE_PURIFY,
+        MOVE_RECOVER,
+        MOVE_REST,
+        MOVE_ROOST,
+        MOVE_SHORE_UP,
+        MOVE_SLACK_OFF,
+        MOVE_SOFT_BOILED,
+        MOVE_STRENGTH_SAP,
+        MOVE_SWALLOW,
+        MOVE_SYNTHESIS,
+        MOVE_WISH,
+    };
+
+    BOOL isTriageMove = FALSE;
+
+    for (u16 i = 0; i < NELEMS(TriageMovesList); i++) {
+        if (TriageMovesList[i] == sp->current_move_index) {
+            isTriageMove = TRUE;
+            break;
+        }
+    }
+
+    if (
+        (sp->moveTbl[sp->current_move_index].priority > 0) ||
+        ((GetBattlerAbility(sp, attacker) == ABILITY_PRANKSTER) &&
+         (sp->moveTbl[sp->current_move_index].split == SPLIT_STATUS) &&
+         (sp->moveTbl[sp->current_move_index].priority >= 0)  // Prankster is +1
+         ) ||
+        ((GetBattlerAbility(sp, attacker) == ABILITY_GALE_WINGS) &&
+         (sp->moveTbl[sp->current_move_index].type == TYPE_FLYING) &&
+         (sp->moveTbl[sp->current_move_index].priority >= 0)  // Gale Wings is +1
+         ) ||
+        ((GetBattlerAbility(sp, attacker) == ABILITY_TRIAGE) &&
+         (isTriageMove) &&
+         (sp->moveTbl[sp->current_move_index].priority >= -2)  // Triage is +3
+         )) {
+        return TRUE;
+    }
+    return FALSE;
+}
+
+/**
+ *  @brief function to check whether a mon is grounded or not
+ *  @param sp global battle structure
+ *  @param client_no resolved battler
+ *  @return `TRUE` if grounded, `FALSE` otherwise
+ */
+BOOL monIsGroundedForPsychicTerrain(struct BattleStruct *sp, u32 client_no) {
+    u8 holdeffect = HeldItemHoldEffectGet(sp, client_no);
+
+    if ((sp->battlemon[client_no].ability != ABILITY_LEVITATE && holdeffect != HOLD_EFFECT_UNGROUND_DESTROYED_ON_HIT  // not holding Air Balloon
+         && (sp->battlemon[client_no].moveeffect.magnetRiseTurns) == 0 && sp->battlemon[client_no].type1 != TYPE_FLYING && sp->battlemon[client_no].type2 != TYPE_FLYING) ||
+        (holdeffect == HOLD_EFFECT_HALVE_SPEED                                     // holding Iron Ball
+         || (sp->battlemon[client_no].effect_of_moves & MOVE_EFFECT_FLAG_INGRAIN)  // is Ingrained
+         || (sp->field_condition & FIELD_STATUS_GRAVITY))) {
+        // not in a semi-vulnerable state
+        if ((sp->battlemon[client_no].effect_of_moves & (MOVE_EFFECT_FLAG_FLYING_IN_AIR | MOVE_EFFECT_FLAG_DIGGING | MOVE_EFFECT_FLAG_IS_DIVING | MOVE_EFFECT_FLAG_SHADOW_FORCE)) == 0) {
+            return TRUE;
+        }
+    }
+
+    return FALSE;
+}
```

Comment:
potentially just do like.  IsClientGrounded or something, it's a general-use func that can be used wherever